### ISSUE TITLE
Есть продвижения?

### ISSUE THREAD

**Issue by manchelsi** [MinimumLaw/TCL-B220G-Linux#1]
Удалось ли запустить Linux на этом железе?

**Comment by MinimumLaw** [comment]
Увы. В связи с сильно возросшей загрузкой проект приостановлен.

В целом ситуация выглядит очень грустно - классическая загрузка с использованием ACPI таблиц оборудования невозможна. Разработчик реализовал самый минимум остальное оставив на откуп дравйерам в WIndows, а получить сколько-нить работающую DTS'ку методом тыка (и, может быть до какой-то степени реверса платы)... Надо быть большим оптимистом, чтоб верить в то что это возможно.

Единственная надежда на какие-то контакты с TCL и их добрую волю. Но... Я бы много на это не ставил. Идея остается, но ресурсов для ее реализации катастрофически не хватает.

**Comment by manchelsi** [comment]
А не знаете ли какие нибудь подобные arm ноутбуки для Linux, есть на рынке?

**Comment by MinimumLaw** [comment]
Да хватает. Даже на том же Qualcom'е

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/arm64/boot/dts/qcom

А дальше смотрим на то, что начинается с sc7180/sc7280. В каком-то виде там будет и поддержка Linux. Как правило это ARM'овские хромбуки. Потому и DTS для них присутствует - его банально сам производитель сделал. Ну и Qualcom'ом можно не ограничиваться.

**Comment by manchelsi** [comment]
Спасибо, значит будем искать хромобук

**Comment by AliceCyber** [comment]
https://github.com/aarch64-laptops/debian-cdimage

https://github.com/merckhung/linux_on_arm64_laptop

https://github.com/hexdump0815/imagebuilder

https://www.youtube.com/watch?v=sMRur1pCW9Q

Ко мне сия цацка уже едет заместо моего семилетнего потаскунчика на атоме. Буду мучать. 

Шансы поднятия mainline дистрибутивов довольно высоки:
- На Snapdragon 7c Gen 1 есть хромбуки у которых можно посмотреть device tree - например, ~HP Chromebook x2 (2021)~ (не идеален: тачпад и клава тут по USB т.к. отделяемые) и Acer Chromebook 511 (C741L/C741LT)
- Snapdragon 7c Gen 2 на котором в видосе выше был запущен линукс - не более чем разогнанный Gen 1, так что образы должны работать одинаково, SoC по сути тот же самый.

### UPD:
- Начинать копать можно со стороны sc7180-trogdor*.dts в [официальных сорцах ядра](https://github.com/torvalds/linux/tree/master/arch/arm64/boot/dts/qcom) и со стороны ChromeOS
- Trogdor - платформа/плата Сhromebook к которой относится как минимум Acer Chromebook 511 (C741L/C741LT)
- Тачпад (и клавиатура?) у платформы/платы [подключен по i2c](https://github.com/torvalds/linux/blob/master/arch/arm64/boot/dts/qcom/sc7180-trogdor-lazor-limozeen-nots-r4.dts) как и в нашем случае.
- черный экран при запуске линя вероятнее всего вызывается [кривой попыткой инициализации драйвера SMMU, что отвечает за I/O к периферии, который не умеет инициализироваться по таблицам ACPI](https://github.com/aarch64-laptops/debian-cdimage/issues/21#issuecomment-1107663672)
- Со стороны ChromeOS нас интересуют конечно же [ядра, ветки chromeos-5.4 и выше](https://chromium.googlesource.com/chromiumos/third_party/kernel/) а также, вероятно, оверлеи [overlay-trogdor](https://chromium.googlesource.com/chromiumos/overlays/board-overlays/+/refs/heads/main/overlay-trogdor/), [baseboard-trogdor](https://chromium.googlesource.com/chromiumos/overlays/board-overlays/+/refs/heads/main/baseboard-trogdor/) и [chipset-qc7180](https://chromium.googlesource.com/chromiumos/overlays/board-overlays/+/refs/heads/main/chipset-qc7180/)

### UPD2 - читая DTS платформы Trogdor наткнулась на интересные вещи:
- Блютуз торчит через UART
- eDP экран сидит через MIPI-DSI->DP преобразователь.
Ничего не напоминает? Да, та же самая песня что и у TLC B220G - скорее всего, в обоих случаях скопирован референсный дизайн платформы на Snapdragon 7c.

**Comment by AliceCyber** [comment]
![IMG_20230520_135700_071](https://github.com/hexdump0815/imagebuilder/assets/45756018/2c90adce-5eae-49e4-8a28-5a0b150fc541)
Три дня до гуя. Приехал он мне 17-го, счас 20-е.
Щас буду поднимать встроенную клаву с тачпадом.
@manchelsi @MinimumLaw ну а у вас дела как?

**Comment by MinimumLaw** [comment]
> @manchelsi @MinimumLaw ну а у вас дела как?

Я рад за вас. Да, все именно так. Рекомендую i2c-detect и с его помощью определится с шиной и чипом тачпада.
Bluetooth строго по стандарту - UART (2 или 4 линии с CTS/RTS) - соответственно H5 и H4, USB, SDIO. Крайне редко SPI.
Мостики из MIPI - довольно распространенная практика. На моем TCL тоже он стоит.

По поводу референсного дизайна... Ну, так или иначе это чистая правда. Но вот что именно сделали TCL с тем референсом.... У многих распаян прямо процессорный модуль с обвязкой. Здесь же распаян чип и обвязка. Что усложняет. Особенно с учетом того, что не только GUI, но и консоли нет. Ни обычной, ни последовательной. Ту же I2C-detect не натравить... 

Но основное - это время. Которого просто нет. Очень много работы и этим надо пользоваться. Потому как пока есть работа есть деньги. А значит есть жизнь. Хобби подождет. Оно не заброшено, но отложено...

**Comment by exstrim401** [comment]
@AliceCyber привет, есть прогресс?

**Comment by sysofd** [comment]
Всем привет, тоже очень интересно. Так как планирую купить данный девайс и хотелось бы пользоваться Linux.
Да и очень странным выбором смотрится Windows для такого девайса.

**Comment by MinimumLaw** [comment]
Боюсь вас расстроить, но... Лучше идите путем @AliceCyber TCL-B220G очень плохой выбор для такого использования.

Если очень кратко, то прогресса нет. Если подробнее, то ситуация выглядит так - пока нет схемы (хотя бы подробной структурной) нет возможности запустить Linux. Таблица устройств (DSTD) в BIOS крайне бедна и недостаточна для работы Linux даже в самом минимальном варианте. А создать DTS (аналог этой самой таблицы DSDT из мира встраиваемых устройств) без схемы очень сложно. Я, если честно, уже перегорел. Изредка, конечно, пытаюсь что-то сделать, но... Проще взять машинку, которая изначально позволяет чем заниматься этим вот мазохизмом.

Более того, там и с Windows не все гладко. У microsoft нет возможности скачать ISO образ для ARM напрямую, но... варианты есть. Так вот - даже если скачать этот образ с помощью системных средств, то Windows переустановить не получится. В штатной поставке не работает не только встроенные клавиатура с по трекпадом, но и подключенные по USB мышь с клавиатурой. Даже для Windows надо интегрировать драйвера в инсталятор.

Решать вам, но пути два. Или покупаете TCL и присоединяетесь ко мне, или идете другим путем.

**Comment by sysofd** [comment]
https://www.tcl.com/global/en/support-mobile/model/tcl-book-14-go - ссылка на драйверы. На всякий случай, так как по мне гуглится она очень плохо.

**Comment by MinimumLaw** [comment]
@sysofd - большое спасибо. К данной теме непосредственно отношения не имеет, но сделать флешку для переустановки Windows позволит. А это, в свою очередь, потенциально позволит чуть более смело обращаться с железкой. А то EDL режим - это хорошо, но... Далеко не панацея.

**Comment by nektovse** [comment]
Заинтересовал данный ноутбук, а не пробовал никто установить сюда Windows 11 IoT Enterprise LTSC 2024 arm64 версию? LTSC версии требуют меньше ресурсов обычно. Хотя бы в качестве эксперимента

**Comment by MinimumLaw** [comment]
Нет. Я предпочел оставить тот вариант Windows, который был на ноутбуке изначально. По крайней мере это исключает проблемы с лицензированием (таблица MSDM в ACPI присутствует). 

В принципе, ничего невозможного нет. Единственная неприятность - штатные образы скорее всего напрямую не заработают по причине отсутствия драйверов. Но можно попробовать с помощью того же UUPDump'а сделать инсталлятор с интегрированными драйверами. Это вполне может сработать.

Мне куда как более интересно увидеть на нем Linux. Но у меня идеи на эту тему, к сожалению, закончились. Жить с ним и Windows можно, но... Лично мне не интересно.

**Comment by fixinit75** [comment]
Изначальная винда плюс одно или два обновления выпуска - лучший вариант, если не выходит поставить линукс. У меня хватило ума поставить 26100 (это 24H2) сюда, так начались проблемы с выходом из сна: тупо часто не реагирует на крышку, надо по тачпаду провести. Переустановка всех дров не помогла. Потом выяснилось, что это на всех новых arm ноутах происходит и Microsoft об этом знают, но чинить не собираются. Последняя же KBшка мне напрочь убила то преимущество, из-за которого я вообще брал этот ноут - автономность в спящем режиме. Он тупо стал то-ли не попадать в него, то-ли самостоятельно выходить из него и не входить обратно. В итоге, я получаю на следующий день разряженный горячий ноут. Такая фигня была раньше только на x86. В общем, единственный вариант - это пытаться ставить сюда линукс, но я слишком тупой и у меня нет времени, чтобы разбираться с DTSками, которые бы тут заработали. Существующие я пробовал, но дошёл только до момента монтирования файловых систем, где я успешно завис из-за того, что usb не инициализируется. Больше у меня идей нет, поэтому если я до середины лета, пока у меня будет свободное время, не смогу запустить тут линукс нормально, то забью и куплю мак. Для моих задач и требований к железу это будет лучший выбор, чем повторно наступать на те же грабли, покупая Snapdragon X ноуты, которые стоят столько же.
Ну а всем остальным советую не обновляться выше 23H2.

**Comment by MinimumLaw** [comment]
Я, к сожалению или к счастью, не пользуюсь им. Ну не интересен он мне под Windows. Потому и специфических заморочек не знаю.

По DTS - так себе путь. Методом тыка - нерельно, схема от производителя - нереально. Пытаться загрузиться с ACPI а не с device-tree - бестолку. Там DSDT тоже никакая. Традиционная тема - вендорская прослойка, которая все правит. Так что для Linux эта машина подходит очень плохо. И да, похоже мак будет лучшим выбором. Я брал по работе с исследовательскими целями. Не более того. Заниматься серьезно им нет ни времени, ни желания. Что сразу попалось - то изложено.

Ну и диск аварийного восстановления был нужен. Раз уж производитель его не приложил. Его я сделал. Мало ли... Да, с помощью UUPDump слил образ, с помощью DISM интегрировал драйвера и проверил в работе. Не помню как именно, но рецепт нашелся довольно быстро. Все. Мне достаточно.

**Comment by nektovse** [comment]
Для инфо: у меня версия 8/256 gb попробовал сделать свой образ интегрировав драйвера DISM из изначально установленной системы, всё получилось.

Microsoft Windows 11 IoT Enterprise/LTSC 2024 version 24H2 build 26100.1742 - Оригинальные образы от Microsoft MSDN [En]
Если кому нужно: https://mega.nz/file/XFkyBArC#W8JyTMwCQsizVYznRDcSQhibUG6ilJHML16lGLwwdxA

Различий в производительности не заметил. В целом делал ради эксперимента и мне LTSC версии win больше нравятся.


Ещё на просторах github нашёл вот такую тему: https://github.com/foric27/kernel_tcl_b220g
Я так понимаю это кастомное ядро Linux, может поможет в чьих то экспериментах с linux.

**Comment by fixinit75** [comment]
> Ещё на просторах github нашёл вот такую тему: https://github.com/foric27/kernel_tcl_b220g Я так понимаю это кастомное ядро Linux, может поможет в чьих то экспериментах с linux.

Я так и не понял, что именно там отличается от оригинального ядра, но обязательно попробую 6.11 скомпилить и подсунуть в Ubuntu live. Да, не лучший способ проверить, но у меня очень мало времени и я почти без выходных пахать буду аж до начала лета.. Если что-то выйдет, напишу комментарий сюда. Спасибо за ссылку! ^ ^

**Comment by MinimumLaw** [comment]
@nektovse - спасибо за проделанную работу  работу. Если не возражаете я укажу ссылку за вашим авторством в README.md

Вопрос про лицензию, все же остался открытым. Насколько комфортно можно пользоваться тем образом? В смысле обычных приколов с нелицензионными версиями? Типа отсутствия обновлений, черного экрана или чего-то подобного.

@fixinit75 - я попробовал связаться с @foric27 У него локация Россия - вполне возможно что получится пообщаться. Но да - в первом приближении DTS нет, defconfig нет. Как собирать не очень понятно. Нужны комментарии автора.

**Comment by nektovse** [comment]
@MinimumLaw 

Да можно указать и без авторства, тут ничего такого нет.
Активируется с помощью https://github.com/massgravel/Microsoft-Activation-Scripts/blob/master/MAS/Separate-Files-Version/Activators/HWID_Activation.cmd

Пользоваться можно, проблем нет, обычная windows, у которой меньше предустановленного софта.  Обновления работают.

**Comment by fixinit75** [comment]
> У него локация Россия - вполне возможно что получится пообщаться.

О, это хорошо. Может, он прояснит.

> DTS нет, defconfig нет

Ну, как я вижу, там defconfig есть дефолтный от arm64. Могу ошибаться, ибо уже ровно год не доводилось компилить ядро под эту платформу, последний раз был, как раз, под этот ноут.
Могу предположить, что можно просто со стандартным его скомпилить и посмотреть, что будет. Хотя, крайне мало в это верю.

**Comment by foric27** [comment]
Тут редко бываю, если что телеграм @foric27

**Comment by MinimumLaw** [comment]
Ну, да там все идет очень интересно. Пока я тут своими другими железками занимался...

https://github.com/hexdump0815/imagebuilder/issues/136#issuecomment-1556219317 - тут набросали DTS, да не просто набросали, а так что даже консоль (вроде бы) работает.

@TravMurav который мне сильно помог, @AliceCyber которая похоже имеет эту игрушку и активно пытается завести. Надо выкраивать время и присоединяться.

**Comment by fixinit75** [comment]
Скомпилил образ отсюда: https://github.com/foric27/kernel_tcl_b220g
С его же defconfigом, без dts при запуске чёрный экран. В общем, не понимаю, что именно должен был дать этот репозиторий. Лучше буду работать с dts от https://github.com/hexdump0815/imagebuilder/issues/136#issuecomment-1556219317

**Comment by MinimumLaw** [comment]
Да, я общался с ним вчера. DTS в репозитарии нет. Со слов автора он добился работоспособности клавиатуры, тача, накопителя, видео и еще какой-то периферии. Разобрали DSDT из BIOS и прописали все найденное в DTS. Сначала руками, потом нейронкой (что бы это не значило). Надо спросить - может и выдаст свой вариант DTS'ки. Без нее грузиться смысла нет вообще.

Хм... а если у нас действительно разные ревизии и в свежих DSDT куда как более богатая? Настолько, что таки можно загрузиться с ACPI вместо DSDT?. Да ну нафиг - бред какой-то (с).

Мои опыты были с левой DTS от ноута на аналогичном чипе (не надо так делать - шанс пожечь шкалит, у меня была мотивация к такому шагу и я понимал ответственность). DTS по ссылке сделана примерно так же - т.е. сдампили и разобрали таблицу DSDT из BIOS и на ее основе создали DTS. Но, судя по всему, она не очень рабочая. В том смысле, что до рабочей консоли не доходит. Даже на отладочном порту. Так что дальше меня ушли, но не на много. По ощущениям скорее нащупали часть периферии. Так вслепую искать - уж очень грустное занятие. Не мудрено, что его бросили.

**Comment by fixinit75** [comment]
Видимо, мой уровень познаний не позволяет мне скомпилить dts...

![Image](https://github.com/user-attachments/assets/c63b6ce1-3b84-404a-9ad1-824cc0e4fd2b)



UPD: Дошло глянуть в качестве референса DTS от sc7180-idp в 6.14 и сравнить со старым от 6.1. DTS несовместим с 6.14, моя ошибка. Может, попробую его подправить
UPD2: скрины с содержимыми референса: первый - 6.14, второй - 6.1

![Image](https://github.com/user-attachments/assets/93eb7b05-8f71-484c-ae4e-f402e5c82b60)

![Image](https://github.com/user-attachments/assets/409d734e-d9bc-4e4b-bb87-829a87206df8)

**Comment by MinimumLaw** [comment]
Судя по всему их переименовали.

https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/tree/arch/arm64/boot/dts/qcom/sc7180.dtsi?h=v6.14.5#n3265 - dsi0 теперь mdss_dsi0 и так далее. В первом приближении просто добавить приставку mdss_

Ну, да вы уже сами...

**Comment by fixinit75** [comment]
> В первом приближении просто добавить приставку mdss_

Да, оно даже скомпилировалось. Сейчас попробую запуститься и посмотреть, что будет

**Comment by MinimumLaw** [comment]
Скорее всего ничего не будет. Мостик DSI там закоментирован (да, судя по всему он и не на той шине). Видимо, если сказать console=отладочный_порт (по ссылке далее), то увидим ровно то же - непонятки с клоком накопителя. Ну если не так дайте знать.

**Comment by fixinit75** [comment]
И всё, в принципе)

![Image](https://github.com/user-attachments/assets/a042e10d-ea7d-4b2b-8ce9-bedc04c3cf3c)

**Comment by MinimumLaw** [comment]
Ну уже круто. Дисплей работает. USB работает, init стартует - т.е. уже не просто ядро в работе. А как флешку готовили? 

Формально, на данный момент нет особого смысла пытаться запустить инсталятор. Надо live-образ разворачивать на флешине (хоть самый минимальный) и для начала грузиться с него.

**Comment by fixinit75** [comment]
> А как флешку готовили?

Мне стыдно сейчас об этом говорить, но на пофигизме и энтузиазме - взял ubuntu 25.04, скачал исходники их ядра в deb пакете, докинул туда изменённый DTS и после компиляции подсунул DTB в образ. В GRUB указал:

`
        set gfxpayload=keep
        linux	/casper/vmlinuz-deb efi=novamap,noruntime clk_ignore_unused pd_ignore_unused  
        initrd	/casper/new-initrd.gz
        devicetree /casper/dtbs/sc7180-tcl-book-14-go.dtb
`

Конечно, так никто не делает, но у меня мало времени, мне через 30 минут ехать на мероприятие, делаю всё на коленке.

UPD: ммм, классно, доигрался

![Image](https://github.com/user-attachments/assets/d0e14e58-5c47-477d-8d8b-6a317699ed3d)

**Comment by MinimumLaw** [comment]
Почему стыдно? Нормальный подход. Совершенно.

Более того - некоторое время назад он не работал совсем. grub от Ubuntu не работал, только от Debian'а. Да еще и требовал переименовать efi файл. Собственно, потому и спрашиваю. Тоже неплохо.

**Comment by fixinit75** [comment]
> Нормальный подход

 Я бы лучше использовал что-то более легковесное, чего у меня нет. В общем, после запуска systemd-udevd ядро виснет намертво. С другими dts у меня при каких-то конфигурациях получалось дойти до безуспешной попытки монтирования файловых систем, но не работало большинство функций, включая клавиатуру. Так что, не могу сказать, насколько улучшилась ситуация. Проверить, что в debug интерфейс печатается, пока не могу...

**Comment by MinimumLaw** [comment]
Ну, за этим проблема не станет - https://rcn-ee.com/rootfs/
По сути любую. А то, что она TI - глубоко наплевать. Все равно ядро свое использовать. А потом apt-ом доставить нужное.

**Comment by fixinit75** [comment]
Чёрт возьми, было бы у меня времени больше... В общем, попробую ещё скомпилировать чистое ядро какое-нибудь, хотя очень хочется от foric27 DTS увидеть, на самом деле.. Если на чистом ядре такая же фигня, как с ubuntu, будет, то даже писать ничего не буду, но хочу попробовать ещё debian. А вообще, мне стало интересно запустить тут FreeBSD, ибо обычно я с ней работаю и сталкиваюсь, но это уже совсем другая тема и история, так что, не буду сюда об этом.

**Comment by fixinit75** [comment]
Кстати, если это может считаться, один раз у меня получилось запустить Fedora на Hyper-V. Это был год 2022-й где-то. И это был единственный дистрибутив, который запустился на этом железе, хоть и в виртуалке.

**Comment by MinimumLaw** [comment]
Я посмотрел кто куда и насколько продвинулся в части запуска Linux.
На сегодня просматривается две ключевые проблемы.

1. Я ядре нет драйвера мостика DSI-eDP LT8911EXB
2. Phy интерфейс UFS накопителя требует нормальной заводки и разбора с его системой тактирования.

Это минимальный объем работ для запуска системы хотя бы в консольном варианте.  Что у нас есть на эту тему?
Есть довольно качественные наброски для драйвера мостика от @aystshen - https://github.com/aystshen/lontium_lt8911exb_driver По сути, его даже можно использовать напрямую. Тайминги дисплея можно взять в логе загрузки uEFI и оттуда прописать в DTS. Но по хорошему, надо бы научить мостик читать EDID от дисплея. Опять же - формально это не сложно. В ядре есть примеры того, как это делать. Но это надо делать и на это надо время. Кстати интересно, а не пробовал автор этот код в ядро включить? Вроде бы вопрос только в документировании DTS. Остальное в первом приближении вполне соответствует принятым правилам. Но да ладно. Я себе в ToDo записал, но далеко не первостепенной задачей к сожалению.

Phy интерфейс пока особенно ничего не скажу. Судя по всему никакого внешнего тактирования там нет, и, если это так, то это вопрос лишь грамотного указания. Возможно для этого придется поиграться с WIndows (найти утилиту и почитать физическую память по разным смещениям). Но, мне кажется, вопрос с видео куда как более приоритетный. не вечно же keep в bootargs'ах держать.

Пока как-то так.

**Comment by Alphizi** [comment]
На маркеплейсах и авито только этот TCL есть из ноутов на snapdragon 7c. Очень хотелось бы Linux, жаль что никому не удалось завести

**Comment by fixinit75** [comment]
Greetings! А у меня, кажется, закончилась история с линуксом, поскольку у моего устройства накрылись кнопки на клавиатуре. Так что, я постепенно стремлюсь к покупке MacBook, ну а всеем вам желаю удачи в дальнейших попытках!) У вас получится, я в вас верю!
